plugins/docker/src/main/java/org/radargun/service/DockerService.java

```diff
@@ -0,0 +1,251 @@
+         .withTailAll()
+         .exec(loggingCallback);
+
+      try {
+         loggingCallback.awaitCompletion(3, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+         e.printStackTrace();
+      }
+      log.infof("Container log:\n%s", loggingCallback.toString());
+   }
+
+   private static class LogCallback extends LogContainerResultCallback {
+      protected final StringBuilder log = new StringBuilder();
+
+      @Override
+      public void onNext(Frame frame) {
+         log.append(new String(frame.getPayload()));
+      }
+
+      @Override
+      public String toString() {
+         return log.toString();
+      }
+   }
+
+   @Override
+   public void stop() {
+      dockerClient.stopContainerCmd(dockerContainerId).exec();
+      started = false;
+      log.infof("Stopped container with id %s", dockerContainerId);
+   }
+
+   @Override
+   public boolean isRunning() {
+      return started;
+   }
+
+   private static class EnvsConverter implements Converter<List<String>> {
+      @Override
+      public List<String> convert(String string, Type type) {
+         List<String> env = new ArrayList<>();
+         String[] lines = string.split("\n");
+         for (String line : lines) {
+            env.add(line.trim());
+         }
+         return env;
+      }
+
+      @Override
+      public String convertToString(List<String> envs) {
+         StringBuilder sb = new StringBuilder();
+         for (String envVar : envs) {
+            sb.append(envVar).append('\n');
+         }
+         return sb.toString();
+      }
+
+      @Override
+      public String allowedPattern(Type type) {
+         return Converter.ANY_MULTI_LINE;
+      }
+   }
+
+   private static class PortsConverter implements Converter<List<ExposedPort>> {
+      @Override
+      public List<ExposedPort> convert(String string, Type type) {
+         List<ExposedPort> ports = new ArrayList<>();
+         String[] portSlashProtocolList = string.trim().split("\n");
+         for (String portSlashProtocol : portSlashProtocolList) {
+            if (portSlashProtocol.split("/").length != 2)
+               throw new IllegalArgumentException("The exposed port must be specified in this format: port/[udp|tcp]");
+            int port = Integer.parseInt(portSlashProtocol.trim().split("/")[0]);
+            InternetProtocol protocol = InternetProtocol.parse(portSlashProtocol.trim().split("/")[1]);
+            ports.add(new ExposedPort(port, protocol));
+         }
+         return ports;
+      }
+
+      @Override
+      public String convertToString(List<ExposedPort> ports) {
+         StringBuilder sb = new StringBuilder();
+         for (ExposedPort port : ports) {
+            sb.append(port.getPort()).append("/").append(port.getProtocol().toString()).append('\n');
+         }
+         return sb.toString();
+      }
+
+      @Override
+      public String allowedPattern(Type type) {
+         return Converter.ANY_MULTI_LINE;
```

Comment:
use proper regexp to match the `port/[udp|tcp]` (btw. I would use round parentheses as `[]` usually means optional part)